'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.snapshotSerializer = undefined;

var _StyleRepresentation = require('./StyleRepresentation');

var snapshotSerializer = exports.snapshotSerializer = {
  // $ExpectError: ...
  print: function print(val, serialize, _indent) {
    if (val instanceof _StyleRepresentation.CSSClassRepresentation) {
      var repr = serialize(val.representation);
      return repr.replace(/^Object {/, val.representationName + ' {');
    } else if (val instanceof _StyleRepresentation.CSSClassJoinRepresentation) {
      var _repr = serialize(val.styles);
      return _repr.replace(/^Array \[/, 'StyleJoin [');
    } else {
      return serialize(val);
    }
  },


  // $ExpectError: ...
  test: function test(val) {
    return val instanceof _StyleRepresentation.CSSClassRepresentation || val instanceof _StyleRepresentation.CSSClassJoinRepresentation;
  }
};
